### PR TITLE
Refactor HeronExecutor signal handling

### PR DESCRIPTION
**Pull request by billonahill**
Cherry picking refactorings that I made during scaling that can be committed first, since they don't relying on the scaling functionality (but will be needed in some cases). No new functionality here, just moving some things around to clean up before adding new functionality.

1. Refactored how exit signal handling is done. This will help when we introduce a state manager that needs closing.
1. Renamed some public methods (`prepare_launch()` -> `initialize()`, `monitor_process()` -> `start_process_monitor()`)
1. Added synchronization around `processes_to_monitor`. 

Relates to #1292

### CODE REVIEW

**Comment by kramasamy** [comment]
looks good to me 👍

File: heron/executor/src/python/heron_executor.py [hunk 1]
```diff
@@ -539,40 +564,32 @@ def get_command_changes(self, current_commands, updated_commands):
...
-    commands = [create_folders, chmod_binaries]
+      # get the commands to kill, keep and start
+      commands_to_kill, commands_to_keep, commands_to_start = \
+          self.get_command_changes(current_commands, updated_commands)
+
+      Log.info("current commands: %s" % sorted(current_commands.keys()))
+      Log.info("new commands    : %s" % sorted(updated_commands.keys()))
+      Log.info("commands_to_kill: %s" % sorted(commands_to_kill.keys()))
+      Log.info("commands_to_keep: %s" % sorted(commands_to_keep.keys()))
+      Log.info("commands_to_start: %s" % sorted(commands_to_start.keys()))
```
Comment (objmagic) [inline heron/executor/src/python/heron_executor.py]: should these be `Log.debug`?

Comment (objmagic) [inline heron/executor/src/python/heron_executor.py]: oh I forgot `heron-executor` logs to a file. Then I think it's ok to do `Log.info` here.

Comment (billonahill) [inline heron/executor/src/python/heron_executor.py]: Yeah, I think this are ok as info. They happen rarely and we typically want to know what's going on with this process restarting part.

**Comment by kramasamy** [comment]
furthermore, these statements are not seen by the user since heron executor is automatically started by the scheduler

**Pull request merged**